src/libgit2/transports/smart_protocol.c

```diff
@@ -339,6 +366,35 @@ int git_smart__negotiate_fetch(git_transport *transport, git_repository *repo, c
 	if ((error = git_revwalk__push_glob(walk, "refs/*", &opts)) < 0)
 		goto on_error;
 
+	if (wants->depth > 0) {
+		git_pkt_shallow *pkt;
+
+		if ((error = git_smart__negotiation_step(&t->parent, data.ptr, data.size)) < 0)
+			goto on_error;
+
+		while ((error = recv_pkt((git_pkt **)&pkt, NULL, buf)) == 0) {
+
+			if (pkt->type == GIT_PKT_SHALLOW) {
+				git_shallowarray_add(wants->shallow_roots, &pkt->oid);
+			} else if (pkt->type == GIT_PKT_UNSHALLOW) {
+				git_shallowarray_remove(wants->shallow_roots, &pkt->oid);
+			} else if (pkt->type == GIT_PKT_FLUSH) {
+				/* Server is done, stop processing shallow oids */
+				break;
+			} else {
+				git_error_set(GIT_ERROR_NET, "Unexpected pkt type");
+				goto on_error;
+			}
+
+			git_pkt_free((git_pkt *) pkt);
+		}
+
+		git_pkt_free((git_pkt *) pkt);
```

Comment:
Does this have the potential for a double-free in the case that the `while` condition is falsey?